runtime/compiler/optimizer/JProfilingValue.cpp

```diff
@@ -454,168 +481,238 @@ TR_JProfilingValue::addProfilingTrees(
    if (roundedType == TR::Address)
       roundedType = systemType;
 
-   if (trace)
-      {
-      traceMsg(comp, "Inserting value profiling trees:\n  Value n%dn At n%dn\n  Table: %p\n",
-         value->getGlobalIndex(),
-         insertionPoint->getNode()->getGlobalIndex(),
-         table);
-      table->dumpInfo(comp->getOutFile());
-      }
-
    TR::Block *originalBlock = insertionPoint->getEnclosingBlock();
-   TR::Block *extendedBlock = originalBlock;
-   while (extendedBlock->isExtensionOfPreviousBlock())
-      extendedBlock = extendedBlock->getPrevBlock();
-
-   if (trace)
-      {
-      traceMsg(comp, "  Modifying block_%d", originalBlock->getNumber());
-      if (extendedBlock && extendedBlock != originalBlock)
-         traceMsg(comp, " extending block_%d", extendedBlock->getNumber());
-      traceMsg(comp, " with profiling\n");
-      }
-
-   // Cache the last treetop in the CFG
    TR::CFG *cfg = comp->getFlowGraph();
    cfg->setStructure(0);
-   TR::Block *cursor = originalBlock, *prev = originalBlock;
-   while (cursor && cursor != cfg->getEnd() && cursor != cfg->getStart() && cursor->getNumber())
+   /**
+    * Anchor the value node after the split point to make sure that it will be either stored in the register or temp slot before split point
+    * Most of the time, a last simplifier run will clean it up but in case it does not, we end up with the unguarded null dereference in case we are
+    * profiling VFT load from object. To avoid that we do not anchor the node when null test is needed.
+    * If the profiling place holder is generated in this pass, we make sure that we generate the NULL Test before actual type test so in that case,
+    * it will be uncommoned by the block splitter. If VFT Profiling is added in other passes, we will store the value in to temp slot for helper call
+    * before we jump to helper call block.
+    */
+   if (!addNullCheck)
+      insertionPoint->insertAfter(TR::TreeTop::create(comp, TR::Node::create(TR::treetop, 1, value)));
+   TR::Block *mainlineReturn = originalBlock->splitPostGRA(insertionPoint->getNextTreeTop(), cfg, true, NULL);
+   
+   TR::Node *origBlockGlRegDeps = originalBlock->getExit()->getNode()->getNumChildren() == 1 ? originalBlock->getExit()->getNode()->getFirstChild() : NULL;
+   if (trace)
+      traceMsg(comp, "\t\t\tUsing split mainline return block = %d\n", mainlineReturn->getNumber());
+   TR::Node *profilingValue = value;
+   /**
```

Comment:
I don't think this should be doxygen

Reply:
Fixed in https://github.com/eclipse/openj9/pull/6585/commits/4babc9149b99cc98d9703d5340c8c21f81c918da